### PR TITLE
Fix runtime_config API semantics and validation

### LINKED ISSUE #3676: Allow runtime image to be specified through UI/API
Following on from #3662 - change the API to allow the runtime image to be passed when creating a workload.

### PR DESCRIPTION
**Pull request by slyt3**
## Summary

Reopening closed PR #3772 and carries over the original `runtime_config` workload API work with the review feedback addressed.

The previous version changed runtime config merge behavior for all callers, including CLI, and allowed some invalid `runtime_config` cases to be ignored or fail later during image build. This update keeps the behavior scoped to the workload API, adds validation, and tightens the response/update round-trip behavior.

Fixes #3676

## Type of change

- [x] Bug fix
- [ ] New feature
- [ ] Refactoring (no behavior change)
- [ ] Dependency update
- [ ] Documentation
- [ ] Other (describe):

## Test plan

- [ ] Unit tests (`task test`)
- [ ] E2E tests (`task test-e2e`)
- [ ] Linting (`task lint-fix`)
- [x] Manual testing (describe below)

Manual testing:
- `go test ./pkg/runner -run 'TestLoadRuntimeConfig_' -count=1`
- `go test ./pkg/api/v1 -run 'Test(RuntimeConfigFromRequest|RuntimeConfigForImageBuild|CreateWorkload|UpdateWorkload)$' -count=1`
- `go test ./pkg/api/v1 -run 'Test(RuntimeConfigForImageBuild|RunConfigToCreateRequest)$' -count=1`

## Changes

| File | Change |
|------|--------|
| `pkg/runner/protocol.go` | Restore runner/CLI replacement semantics for runtime overrides |
| `pkg/runner/protocol_test.go` | Update runtime config tests and assertion style |
| `pkg/api/v1/workload_service.go` | Add API-scoped merge and validation for `runtime_config` |
| `pkg/api/v1/workload_service_test.go` | Add validation and protocol-image edge-case coverage |
| `pkg/api/v1/workload_types.go` | Omit `runtime_config` from non-protocol response payloads |
| `pkg/api/v1/workloads_test.go` | Add create/update coverage for `runtime_config` behavior |
| `pkg/api/v1/workloads_types_test.go` | Add response conversion coverage for protocol vs non-protocol images |
| `docs/server/docs.go` | Update generated API docs |
| `docs/server/swagger.json` | Update generated API docs |
| `docs/server/swagger.yaml` | Update generated API docs |

## Does this introduce a user-facing change?

Yes. Invalid runtime_config input now fails early with a clear API validation error, and runtime_config is only accepted on create/update when image is a protocol URI.

## Special notes for reviewers

This is intended as the replacement for #3772, not a separate feature. The main goal is to keep the runtime-config behavior scoped to the workload API without changing shared runner/CLI semantics.

### CODE REVIEW

**Comment by JAORMX** [comment]
The merge conflicts must be solved.

**Comment by slyt3** [comment]
@JAORMX Yeah I know I was rebasing and resolving conflicts; branch is now cleanly mergeable. I’m fixing the remaining docgen check now.

**Comment by slyt3** [comment]
@JAORMX Hi! This PR should be ready, one E2E test is failing but seems unrelated; I’d really appreciate your review when you have time, thank you :)

**Comment by JAORMX** [comment]
@slyt3 I'll be unavailable this week as I'm flying to MCP Dev Summit. Perhaps you can reach out in Discord and another maintainer can check this out?

**Comment by slyt3** [comment]
@JAORMX Got it, thanks for letting me know! Safe travels to the MCP Dev Summit. I will reach out on Discord and see if another maintainer can take a look.

File: pkg/api/v1/workload_types.go [hunk 1]
```diff
@@ -40,6 +41,8 @@ type workloadStatusResponse struct {
 type updateRequest struct {
 	// Docker image to use
 	Image string `json:"image"`
+	// RuntimeConfig allows overriding runtime build configuration for protocol schemes.
```
Comment (jerm-dro) [inline pkg/api/v1/workload_types.go]: **question:** How will Studio (or other API clients) use `runtime_config`?

There's a potential round-trip issue (also flagged by Codex in [this comment](https://github.com/stacklok/toolhive/pull/4435#discussion_r3008378619)): after a protocol-scheme image is built, `RunConfig.Image` stores the built name (e.g., `toolhivelocal/go-server:...`), not the original `go://...` URI. GET responses return this built name alongside `runtime_config`. On PUT, `runtimeConfigForImageBuild` rejects `runtime_config` when the image isn't a protocol scheme — so a naive GET → modify → PUT round-trip would 400.

Whether this is actually a problem depends on how the API is consumed. If Studio always re-specifies the original `go://...` URI on updates, it's fine. If it echoes back the GET response, it breaks.

Either way, the API contract for `runtime_config` should be documented — specifically that it's only accepted on create/update when the `image` field is a protocol-scheme URI, and that clients should not echo it back from GET responses alongside a built image name. The swagger docs should make this clear so Studio (and future clients) know what to expect.

File: pkg/api/v1/workload_service.go
Comment (jerm-dro) [inline pkg/api/v1/workload_service.go:1]: **blocker:** I noticed several changes that seem unrelated to the `runtime_config` feature — I'm less familiar with this part of the codebase, so they make the review harder to follow:

- `registryProxyPort` logic removed entirely — this drops the registry-provided proxy port override, which looks like a behavioral change
- `imageMetadata` nil guard relaxed — `ok && md != nil` → bare type assertion
- `remoteServerMetadata` nil check relaxed — `ok && remoteServerMetadata != nil` → `ok`
- Transport type fallback changed from `ImageMetadata`-specific to generic `serverMetadata.GetTransport()`
- Flaky test fix in `pkg/vmcp/composer/state_store_test.go`

Are these required for `runtime_config` to work? If they're independent improvements, could you split them into a separate PR? It would make it easier to review the `runtime_config` changes in isolation and attribute any regressions.

File: pkg/api/v1/workload_types.go (hunk 1 above)
Comment (slyt3) [reply pkg/api/v1/workload_types.go]: I added runtime_config so clients can see what was used, but you’re right this can break a simple GET -> edit -> PUT flow after the image is converted to a built image name. I will update the docs/swagger to make it clear: runtime_config is only accepted on create/update when image is a protocol URI (go://, npx://, uvx://). I will also follow up in this PR so Studio/client behavior is clear and this round trip isnt confusing

File: pkg/api/v1/workload_service.go
Comment (slyt3) [reply pkg/api/v1/workload_service.go:1]: those got mixed in while i was rebasing and fixing follow up issues, they are not needed for runtim_config, so I will remove them from this PR

**Comment by slyt3** [comment]
@jerm-dro i removed the unrelated changes from that PR and clarified the runtime_config API contract in the docs/comments

**Comment by jhrozek** [comment]
@jerm-dro were you going to merge the PR? I don't think the contributor can on their own.

**Comment by slyt3** [comment]
@jerm-dro @jhrozek yeah, I cant merge it, just a contributor
